nav2_map_server/README.md

```diff
@@ -1,59 +1,133 @@
 
-In addition, there is now two new "load_map" and "save_map" services which can be used to
-dynamically load and save a map.
+In addition, there is now new "load_map/load_map_3d" and "save_map/save_map_3d" services which can be used to dynamically load and save a map.
 
 ### Architecture
 
-In contrast to the ROS1 navigation map server, the nav2 map server will support a variety
-of map types, and thus some aspects of the original code have been refactored to support
-this new extensible framework.
+In contrast to the ROS1 navigation map server, the nav2 map server will support a variety of map types, and thus some aspects of the original code have been refactored to support this new extensible framework.
 
-Currently map server divides into tree parts:
+#### Extensible Framwork
 
-- `map_server`
-- `map_saver`
-- `map_io` library
+The Nav2 Map Server, leverages *template metaprogramming* to enable handling of new map-types via *template specializations*. This provides switching between different server types via specifying map-type as template parameter. Thus all specializations of map-server share the same `LifecycleNode` structure, while giving to user a complete  access to the design. This approach includes integration of a map I/O system which is responsible for basic operations with any type of map, and server parts for handling services, messages etc. with these maps. It allows to have an isolated per each map-type implementations of the complete map handling system.
 
-`map_server` is responsible for loading the map from a file through command-line interface
-or by using serice requests.
+Demonstration of *template metaprogramming* and *specializations*:
+```c++
+template<class mapT>
+class ExampleClass {
+};
 
-`map_saver` saves the map into a file. Like `map_server`, it has an ability to save the map from
-command-line or by calling a service.
+template<>
+class ExampleClass<specialization_1> {
+};
 
-`map_io` - is a map input-output library. The library is designed to be an object-independent
-in order to allow easily save/load map from external code just by calling necessary function.
-This library is also used by `map_loader` and `map_saver` to work. Currently it contains
-OccupancyGrid saving/loading functions moved from the rest part of map server code.
-It is designed to be replaceble for a new IO library (e.g. for library with new map encoding
-method or any other library supporting costmaps, multifloor maps, etc...).
+template<>
+class ExampleClass<specialization_2> {
+};
+```
 
-### CLI-usage
 
-#### Map Server
+In order to get a new implementation for other map-type with the new map-server, one needs to define a specialization of the base server and saver with the required services/messages types, then
+add the declaration of this specialization into `map_server.hpp` and `map_saver.hpp` interface headers.
```

Comment:
After this line "Extensible Framework" sub-section will end. So, please separate this by adding a new sub-section, say:
#### Components Architecture
.. or anything you may better to invent.